### PR TITLE
'where' standard filter implementation - resolves #374

### LINKED ISSUE #374: Filter "where" not implemented
### Dotliquid version

2.0.333

### Expected behavior

I should be able to use the `where` filter, that is available and documented [HERE](https://shopify.github.io/liquid/filters/where/)

### Actual behavior

`where` filter not available

### Steps to reproduce the Problem (you can add files)

Just try to do a template with `where` filter in it.

I can try to provide a PR with the implementation for evaluation.

### PR DESCRIPTION
**Pull request by gimmi**
resolves #374

### CODE REVIEW

File: src/DotLiquid/WhereFilter.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                null => new object[0],
+                string _ => new [] {any},
+                IDictionary<string, object> dict => dict.Values,
+                IEnumerable<object> en => en,
+                IEnumerable en => en.Cast<object>(),
+                _ => new[] {any}
+            };
+        }
+
+        private static object GetFieldVal(object any, string key)
```
Comment (daviburg) [inline src/DotLiquid/WhereFilter.cs:30]: Avoid non standard acronyms (such as 'Val', 'el', 'dict'), spell out words.

File: src/DotLiquid/WhereFilter.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        public static IEnumerable<object> Where(object any, string key, object value)
+        {
+            return ToEnum(any)
+                .Where(el => Comparer.Equals(GetFieldVal(el, key), value));
+        }
+
+        private static IEnumerable<object> ToEnum(object any)
+        {
+            return any switch {
```
Comment (daviburg) [inline src/DotLiquid/WhereFilter.cs:20]: There is some non obvious logic as to why these conversions need to be. Please add code comments and/or specification reference to explain.

File: src/DotLiquid/WhereFilter.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            }
+
+            return null;
+        }
+
+        private class ValueEqualityComparer : IEqualityComparer<object>
+        {
+            bool IEqualityComparer<object>.Equals(object left, object right)
+            {
+                // This is the same as DotLiquid.Condition.EqualVariables
```
Comment (daviburg) [inline src/DotLiquid/WhereFilter.cs:44]: Then let's move the same code to a share file and not copy it.

File: src/DotLiquid/DotLiquid.csproj [hunk 4]
```diff
@@ -30,6 +30,7 @@
     <DebugType>full</DebugType>
     <DebugSymbols>True</DebugSymbols>
       <DocumentationFile>bin\Debug\$(TargetFramework)\DotLiquid.xml</DocumentationFile>
+    <LangVersion>8</LangVersion>
```
[the remark this answers is not in the document]
Comment (gimmi) [reply src/DotLiquid/DotLiquid.csproj:33]: Already tried, `latest` has the same result. The requirement for C# 8 basically is only for the switch statement, but anyway maybe updating CI build is more appropriate

File: .gitignore [hunk 5]
```diff
@@ -43,4 +43,6 @@ build/
  
 # Exclude ALL DLLs? (
 # *.dll
-*.coverage
\ No newline at end of file
+*.coverage
+
+/src/.idea/
```
[the remark this answers is not in the document]
Comment (gimmi) [reply .gitignore:48]: Is for anyone that use Rider, so yes.

**Comment by gimmi** [comment]
Please note that the `where` filter has been ported from [DotLiquid.Extras](https://github.com/gimmi/DotLiquid.Extras) project.

If the implementation is OK to be merged, then it is a matter of doing some refactoring to make the filter more in line with the codebase.

File: src/DotLiquid/DotLiquid.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (daviburg) [reply src/DotLiquid/DotLiquid.csproj:33]: The highest version we can use is C# 7.3
Version 8 or higher would be incompatible with .NET Framework we must support for our client.
If you have code depending on C# 8 you need to revert it to a 7.3-compatible form.

See
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version

**Comment by daviburg** [comment]
Alternative implementation of the same Where filter has been completed by another author, so closing this PR.

**Pull request closed without merging**